runtime/strategies/coalesce-recipes.js

```diff
@@ -209,9 +209,39 @@ export class CoalesceRecipes extends Strategy {
         return results;
       }
 
-      _connectOtherHandles(otherToHandle, cloneMap) {
-        otherToHandle.forEach((otherHandle, handle) => cloneMap.get(otherHandle).mergeInto(handle));
+      _connectOtherHandles(otherToHandle, cloneMap, verifyTypes) {
+        otherToHandle.forEach((otherHandle, handle) => {
+          let otherHandleClone = cloneMap.get(otherHandle);
+
+          // For coalescing that was triggered by handle coalescing (vs slot or slot connection)
+          // once the main handle (one that triggered coalescing) was coalesced, types may have changed.
+          // Need to verify all the type information for the "other" coalescable handles is still valid.
+
+          // TODO(mmandlis): This is relying on only ever considering a single "other" handles to coalesce,
+          // so the handle either is still a valid match or not.
+          // In order to do it right for multiple handles, we need to try ALL handles,
+          // then fallback to all valid N-1 combinations, then N-2 etc.
+          if (verifyTypes) {
+            if (!this._reverifyHandleTypes(handle, otherHandleClone)) {
+              return;
+            }
+          }
+
+          otherHandleClone.mergeInto(handle);
+        });
       }
+
+      // Returns true, if both handles have types that can be coalesced.
+      _reverifyHandleTypes(handle, otherHandle) {
+        assert(handle.recipe == otherHandle.recipe);
+        let cloneMap = new Map();
+        let recipeClone = handle.recipe.clone(cloneMap);
+        recipeClone.normalize();
+        let handleClone = cloneMap.get(handle);
+        return Handle.effectiveType(
+            handleClone.type, [...handleClone.connections, ...cloneMap.get(otherHandle).connections]);
```

Comment:
might be clearer to roll cloneMap.get(handle).connections in here rather than separate it out too

Reply:
i originally kept it because it's used twice, but it wasn't significantly shorter or more readable. so - done!